napari/utils/colormaps/colormap.py

```diff
@@ -172,67 +162,96 @@ class Config:
     @cached_property
     def _uint8_colors(self) -> np.ndarray:
         data = np.arange(256, dtype=np.uint8)
-        return self.map(data, apply_selection=False)
+        return self._map_without_cache(data)
 
     @cached_property
     def _uint16_colors(self) -> np.ndarray:
         data = np.arange(65536, dtype=np.uint16)
-        return self.map(data, apply_selection=False)
+        return self._map_without_cache(data)
 
-    def selection_as_type(self, dtype: np.dtype) -> int:
-        """Convert the selection value to a specified data type.
+    @cached_property
+    def _int8_colors(self) -> np.ndarray:
+        data = np.arange(256, dtype=np.uint8).astype(np.int8)
+        return self._map_without_cache(data)
+
+    @cached_property
+    def _int16_colors(self) -> np.ndarray:
+        data = np.arange(65536, dtype=np.uint16).astype(np.int16)
+        return self._map_without_cache(data)
 
-        This maps negative background values in int8 and int16 to their
-        corresponding view in uint8 and uint16.
+    def _cmap_without_selection(self) -> "LabelColormapBase":
+        if self.use_selection:
+            cmap = self.__class__(**self.dict())
+            cmap.use_selection = False
+            return cmap
+        return self
+
+    def _get_from_cache(self, values: np.ndarray) -> Optional[np.ndarray]:
+        if values.dtype == np.uint8:
+            return self._uint8_colors[values]
+        if values.dtype == np.uint16:
+            return self._uint16_colors[values]
+        if values.dtype == np.int8:
+            return self._int8_colors[values]
+        if values.dtype == np.int16:
+            return self._int16_colors[values]
+        return None
```

Comment:
You might not like this, and I'm not sure I like it either, but the obvious if/return boilerplate here makes me want a better solution, here's what I've come up with:

```suggestion
        if values.dtype.itemsize <= 2:
            prop_name = '_' + str(values.dtype.type) + '_colors'
            return getattr(prop_name)[values]
        return None
```

Take it or leave it, or maybe think of a nicer way...

Reply:
My preferred solution is to have cache dict. But the problem is that we use EventedModel.
My prefered solution is to refactor Colormaps to not use EvventedModel. Only left few properties (like selected_num) as editable and emitting signal. 

This solution is uglier than your suggestion but better works with "Search usages" utility in IDE, so simplify refactor.

Reply:
Ok. I have used a private attr for cache, so remove this repetitive code.